lib/metasploit/framework/login_scanner/base.rb

```diff
@@ -251,6 +251,8 @@ def scan!
                   break if total_error_count >= 10
                 end
               end
+            rescue => e
+              elog('Attempt may not yield a result', error: e)
```

Comment:
I think this might be too broad of a catch; Is it meant to be just around the `attempt_login` functionality?

It feels like the scanner implementations should be updated to not leak exceptions instead of this `rescue` here 🤔

Reply:
You are correct it is meant to wrap the `attempt_login` and scanners should do a good job of catching exception, in the current form I believe this is pretty much equivalent to:
``` Ruby
              begin
                result = attempt_login(credential)
              rescue => e
                elog('Attempt may not yield a result', error: e)
                next
             end
```
With the minor added rescue of possible errors when post processing the result.

The change is very broad and I think may be possible to improve over time, however as seen in the Postgesql update in this PR there are various errors that occur rarely and as is will bubble up and kill the whole `scan!`.  This `rescue` is meant to provide both better resilience and logging to add a bit of an audit trail we can follow for when things do go sideways.  I was looking for a more limited pattern to catch and did not come up with anything that handles `Errno::*` exceptions well for all supported OS versions as they are not as strongly defined as other socket and timeout related errors.

Reply:
I think I would be in favor of moving the exception handling logic into the postgres attempt_login implementation itself, as this broad error handling might hide issues that should be bubbled up to the user

Reply:
A middleground might be using `Metasploit::Model::Login::Status::UNABLE_TO_CONNECT` here as a result instead of nil, or a introduce a new value to handle the edgecase of `attempt_login` leaking unknown exceptions - it might be a better UX.

Reply:
We could catch all here and raise if it is not a `SystemCallError` but not sure if that really adds much for the user/developer, the real value for me is logging the traces without `crashing` what is often a very long running task.

Reply:
It looks like we miss the `consecutive_error_count` logic handling with this pattern, I'm not sure if that's intentional or not

Maybe something similar to this?

```ruby
begin
  result = attempt_login(credential)
rescue => e
  elog('failed attempting to login', error: e)
  result = {
     credential: credential,
     host: host,
     port: port,
     status: LOGIN_STATUS::UNABLE_TO_CONNECT,
     proof: e.message
  }
end

...
```

Although I'm not sure if protocol is a required field

Reply:
> A middleground might be using `Metasploit::Model::Login::Status::UNABLE_TO_CONNECT` here as a result instead of nil, or a introduce a new value to handle the edgecase of `attempt_login` leaking unknown exceptions - it might be a better UX.

I thought this function wasn't meant to return anything? The YARD docs have it listed as returning VOID.

Reply:
@jmartin-r7  I think adding a new `LOGIN_STATUS::UNKNOWN` status is probably safest, and incrementing the error count logic, since this is a pretty obscure catch all/fallback scenario that we want to handle